tests/test_ssa_generation.py

```diff
@@ -0,0 +1,942 @@
+            for successor in first.sons:
+                if successor not in visited:
+                    worklist.append(successor)
+        return False
+
+    for node in f.nodes:
+        if node is f.entry_point:
+            assert node.immediate_dominator is None
+        else:
+            assert node.immediate_dominator is not None
+            assert find_path(node.immediate_dominator, node)
+
+
+def phi_values_inserted(f: Function):
+    """Verifies that phi-values are inserted at the right places
+
+    For every node that has a dominance frontier, any def (including
+    phi) should be a phi function in its dominance frontier
+    """
+
+    def have_phi_for_var(node: Node, var):
+        """Checks if a node has a phi-instruction for var
+
+        The ssa version would ideally be checked, but then
+        more data flow analysis would be needed, for cases
+        where a new def for var is introduced before reaching
+        DF
+        """
+        non_ssa = var.non_ssa_version
+        for ssa in node.irs_ssa:
+            if isinstance(ssa, Phi):
+                if non_ssa in map(lambda ssa_var: ssa_var.non_ssa_version, ssa.read):
+                    return True
+        return False
+
+    for node in filter(lambda n: n.dominance_frontier, f.nodes):
+        for df in node.dominance_frontier:
+            for ssa in node.irs_ssa:
+                if isinstance(ssa, OperationWithLValue):
+                    if is_used_later(node, ssa.lvalue):
+                        assert have_phi_for_var(df, ssa.lvalue)
+
+
+@contextmanager
+def select_solc_version(version: Optional[str]):
+    """Selects solc version to use for running tests.
+
+    If no version is provided, latest is used."""
+    # If no solc_version selected just use the latest avail
+    if not version:
+        # This sorts the versions numerically
+        vers = sorted(
+            map(
+                lambda x: (int(x[0]), int(x[1]), int(x[2])),
+                map(lambda x: x.split(".", 3), solc_select.installed_versions()),
+            )
+        )
+        ver = list(vers)[-1]
+        version = ".".join(map(str, ver))
+    env = dict(os.environ)
+    env_restore = dict(env)
+    env["SOLC_VERSION"] = version
+    os.environ.clear()
+    os.environ.update(env)
+
+    yield version
+
+    os.environ.clear()
+    os.environ.update(env_restore)
+
+
+@contextmanager
+def slither_from_source(source_code: str, solc_version: Optional[str] = None):
+    """Yields a Slither instance using source_code string and solc_version
+
+    Creates a temporary file and changes the solc-version temporary to solc_version.
+    """
+
+    def build_filename(i: int) -> pathlib.Path:
+        return SCRIPT_DIR / f"slither_test_{i}.sol"
+
+    # NOTE (hbrodin): There is a race condition here (TOCTOU), ignore that for now (only applies to test code).
+    i = 0
+    fname = build_filename(i)
+    while fname.exists():
+        i += 1
+        fname = build_filename(i)
```

Comment:
You can probably still use `NamedTemporaryFile` on Windows if you want, by passing `delete=False`, closing after writing, and then unlinking manually when you are finished.

Reply:
I agree. I'll try that. It would make the could much more clear.